### PR TITLE
dbeaver/dbeaver#34796 Fix table alias for new autocompletion

### LINKED ISSUE #34796: Table alias isn't added with the new autocompletion
### Description

As in the title

### DBeaver Version

24.1.3

### Operating System

Any

### Database and driver

Any

### Steps to reproduce

1. Set preference to add alias for table names in autocompletion
2. Start typing the query
3. On table name select one of the proposals

**Expected:** Table name is added with alias
**Actual:** Alias is not added

### Additional context

https://github.com/user-attachments/assets/67087a44-7b06-4020-8b7d-d16852256109

### PR DESCRIPTION
**Pull request by HocKu7**

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java [hunk 1]
```diff
@@ -329,7 +329,7 @@ private static SyntaxInspectionResult inspectAbstractSyntaxAtState(@NotNull List
             expectingTableName,
             expectingColumnReference,
             reachabilityTests.get(SQLStandardParser.RULE_identifier) || presenceTests.get(SQLStandardParser.RULE_identifier),
-            expectingTableName && reachabilityTests.get(SQLStandardParser.RULE_nonjoinedTableReference),
+            expectingTableName || reachabilityTests.get(SQLStandardParser.RULE_nonjoinedTableReference),
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java]:
```suggestion
            expectingTableName && (reachabilityTests.get(SQLStandardParser.RULE_nonjoinedTableReference) || presenceTests.get(SQLStandardParser.RULE_nonjoinedTableReference)),
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/completion/SQLQueryCompletionContext.java [hunk 2]
```diff
@@ -342,8 +344,16 @@ private List<SQLQueryCompletionItem> prepareObjectComponentCompletions(
...
+                    return components.filter(a -> (componentType == null || componentType.isInstance(a))
+                            && a.getName().toLowerCase().contains(componentNamePart))
+                        .map(x -> {
+                            //todo transform to the map strategy if add one more case
+                            if (x instanceof DBSEntity dbsEntity) {
+                                return SQLQueryCompletionItem.forRealTable(dbsEntity, true);
+                            } else {
+                                return SQLQueryCompletionItem.forDbObject(x);
+                            }
+                        })
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/completion/SQLQueryCompletionContext.java]: Let's move it to the caller, like it's done in `collectObjectRecursively` by adding `Function<T, SQLQueryCompletionItem> completionItemFabric` parameter.
```suggestion
                    return components.filter(a -> (componentType == null || componentType.isInstance(a))
                            && a.getName().toLowerCase().contains(componentNamePart))
                        .map(c -> completionItemFabric.apply((T) c))
```

File: plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/completion/SQLQueryCompletionContext.java [hunk 3]
```diff
@@ -342,8 +344,16 @@ private List<SQLQueryCompletionItem> prepareObjectComponentCompletions(
...
                     }
 
-                    return components.filter(a -> (componentType == null || componentType.isInstance(a)) && a.getName().toLowerCase().contains(componentNamePart))
-                        .map(SQLQueryCompletionItem::forDbObject)
+                    return components.filter(a -> (componentType == null || componentType.isInstance(a))
+                            && a.getName().toLowerCase().contains(componentNamePart))
+                        .map(x -> {
+                            //todo transform to the map strategy if add one more case
+                            if (x instanceof DBSEntity dbsEntity) {
+                                return SQLQueryCompletionItem.forRealTable(dbsEntity, true);
```
Comment (E1izabeth) [inline plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/semantics/completion/SQLQueryCompletionContext.java]: We can't just put true to the second argument. We need to check if it's a first usage of the table or not. Use `alreadyReferencedTables` set for that

**Review by E1izabeth — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java (hunk 1 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model.lsm/src/org/jkiss/dbeaver/model/stm/LSMInspections.java]: And add `SQLStandardParser.RULE_nonjoinedTableReference` to the `presenceTestRule`

**Review by E1izabeth — APPROVED** [review]

**Review by ShadelessFox — APPROVED** [review]
I have no idea what's going on there, but I trust Elizabeth.

**Comment by Matvey16** [comment]
Verified

**Review by serge-rider — APPROVED** [review]

**Pull request merged**